### PR TITLE
Separate dockerfile for Arm compilers

### LINKED ISSUE #94: Use a separate Docker image for proprietary compilers
We currently use one Docker image per Ubuntu version that we're using. Each image has everything needed to run a full Linux CI campaign except for tools that aren't (easily) available on that particular Ubuntu vintage, including Arm proprietary compilers.

The Docker files download Arm compiler binaries. Until recently, those were available to download (but would not run without a license). But now the download is behind a login wall and we aren't supposed to redistribute the binaries. So no one can build our Docker files anymore (not even Arm employees). Even if we have a solution for the CI, it won't be available for people who want to build the Docker image locally.

So I propose to remove the Arm compilers from the `ubuntu-NNN` images (which everyone can build), and have one separate image with the Arm compilers (which will be restricted). This way, problems related to Arm compilers won't interfere with other maintenance of the Docker files.

Task description:

* Copy `ubuntu-20.04/Dockerfile` to make a new image `arm-compilers/Dockerfile`.
* Remove unneeded things from `arm-compilers/Dockerfile`. Do not change anything before the Arm compiler downloads, so that the images can be rebuilt from the existing cached steps.
* Remove the installation of Arm proprietary compilers from `ubuntu*/Dockerfile`.
* Update `linux_platforms` in `vars/common.groovy`.

### PR DESCRIPTION
**Pull request by gilles-peskine-arm**
There are no longer any public URLs to download Arm compilers, so the Docker images can't be built from the dockerfiles anymore. Separate one Docker image with the Arm compilers, and other Docker images with only publicly available tools that anyone can build. Fix #94.

Note that we're relying on cached images for the `arm-compilers` image. Making an image that can be built from scratch if you have access to the required programs is out of scope of this pull request.

Test jobs at 24882ca75c692d9cc3f13cdbaaddb39df4d3dd7b:

* [OpenCI Dockerfile build](https://ci.staging.trustedfirmware.org/job/mbedtls-release-ci-testing/7/) → OK, built the `arm-compilers` image (using cached data) and the new `ubuntu-NN` images
* [OpenCI, `development`, all Linux testing](https://ci.staging.trustedfirmware.org/job/mbedtls-release-ci-testing/8/) → failed `build_armcc`, otherwise passed
* [OpenCI, `mbedtls-2.28`, all Linux testing](https://ci.staging.trustedfirmware.org/job/mbedtls-release-ci-testing/9/) failed `build_armcc`, otherwise passed
* [Internal CI, pr-merge (with ABI check failure)](https://jenkins-mbedtls.oss.arm.com/job/mbed-tls-pr-ci-testing/view/change-requests/job/PR-758-merge/10/) → ran as expected (correct failures reported)
* [Internal CI, `development`, all Linux testing](https://jenkins-mbedtls.oss.arm.com/job/mbedtls-release-ci-testing/434/) failed `build_armcc`, otherwise passed

The failures on `build_armcc` is because the Groovy code relies on `all.sh` to determine whether each component can run, but [we aren't doing that for `build_armcc`](https://github.com/Mbed-TLS/mbedtls-test/pull/96#pullrequestreview-1312737885), so this PR adds a workaround until the [better fix](https://github.com/Mbed-TLS/mbedtls/pull/7163) has spread to all branches we care about.

Test jobs at 87c571485ca83119409e1fa78afa5bc7fd31373b:
* [OpenCI, `development`, `all.sh`](https://ci.staging.trustedfirmware.org/job/mbedtls-release-ci-testing/12/) → passed, the logs of `build_armcc` look fine.

### CODE REVIEW

File: vars/gen_jobs.groovy [hunk 1]
```diff
@@ -609,6 +609,15 @@ aws ecr get-login-password | docker login --username AWS --password-stdin $commo
 """
                         }
 
+                        /* Hack alert: at the time of writing, building the
+                         * arm-compilers image doesn't work, because the
+                         * URLS to download Arm compilers aren't valid anymore.
+                         * So arrange to look for cached copies of the image.
+                         */
+                        if (platform == 'arm-compilers') {
+                            extra_build_args += " --cache-from $common.docker_repoubuntu-20.04-f15359b0ac46e767133991b76ccf23e3cd2e5e0f: "
```
Comment (bensze01) [inline vars/gen_jobs.groovy:618]:
```suggestion
                            extra_build_args += " --cache-from $common.docker_repo:ubuntu-20.04-f15359b0ac46e767133991b76ccf23e3cd2e5e0f"

```
Typo.

**Review by bensze01 — CHANGES_REQUESTED** [review]

**Review by lpy4105 — COMMENTED** [review]
The change here looks fine to me. But since `all_u16-build_armcc` failed on both `development` and `mbedtls-2.28` branch, I created https://github.com/Mbed-TLS/mbedtls/pull/7163 to resolve the failure, and it should be merged before this PR.

**Comment by gilles-peskine-arm** [comment]
Thanks for analyzing the failure @lpy4105. Unfortunately, if we merge this as is, the armcc job will fail on all PR that don't have #7163 merged. So I'll make a workaround in the groovy code.

**Comment by lpy4105** [comment]
It seems that the script didn't generate the job for `build_armcc` correctly as it still failed. I'm not very sure, but at least `node_label_for_platform` and `platform_has_docker` doesn't work correctly with the platform name `arm-compilers`

**Comment by gilles-peskine-arm** [comment]
As I write, the CI is still running, but the `build_armcc` job pas passed so I think this is ready for review.

**Review by bensze01 — APPROVED** [review]
Looks good to me.

**Review by lpy4105 — APPROVED** [review]
LGTM

**Pull request merged**